ddtrace/internal/wrapping.py

```diff
@@ -370,25 +375,50 @@ def wrap(f, wrapper):
     code.argnames = f.__code__.co_varnames[:nargs]
 
     f.__code__ = code.to_code()
-    wf = cast(WrapperFunction, f)
+
+    # DEV: Multiple wrapping is implemented as a singly-linked list via the
+    # __dd_wrapped__ attribute.
+    wf = cast(WrappedFunction, f)
     wf.__dd_wrapped__ = wrapped
 
     return wf
 
 
-def unwrap(wrapper):
-    # type: (WrapperFunction) -> FunctionType
+def unwrap(wf, wrapper):
+    # type: (WrappedFunction, Wrapper) -> FunctionType
     """Unwrap a wrapped function.
 
-    This is the reverse of :func:`wrap`.
+    This is the reverse of :func:`wrap`. In case of multiple wrapping layers,
+    this will unwrap the one that uses ``wrapper``. If the function was not
+    wrapped with ``wrapper``, it will return the first argument.
     """
+    # DEV: Multiple wrapping layers are singly-linked via __dd_wrapped__. When
+    # we find the layer that needs to be removed we also have to ensure that we
+    # update the link at the deletion site if there is a non-empty tail.
     try:
-        wrapped = cast(FunctionType, wrapper.__dd_wrapped__)
-        assert wrapped, "Wrapper has wrapped function"
-        if wrapped.__name__ == "<wrapped>":
-            f = cast(FunctionType, wrapper)
-            f.__code__ = wrapped.__code__
-            del wrapper.__dd_wrapped__
-    except (IndexError, AttributeError):
-        return cast(FunctionType, wrapper)
-    return f
+        wrapped = cast(FunctionType, wf.__dd_wrapped__)
+
+        # Sanity check
+        assert wrapped.__name__ == "<wrapped>", "Wrapper has wrapped function"
+
+        if wrapper not in cast(FunctionType, wf).__code__.co_consts:
+            # This is not the correct wrapping layer. Try with the next one.
+            return unwrap(cast(WrappedFunction, wrapped), wrapper)
+
+        # Remove the current wrapping layer by moving the next one over the
+        # current one.
+        f = cast(FunctionType, wf)
+        f.__code__ = wrapped.__code__
+        try:
+            # Update the link to the next layer.
+            wf.__dd_wrapped__ = cast(WrappedFunction, wrapped).__dd_wrapped__  # type: ignore[assignment]
```

Comment:
can we fix this type ignore? we're already casting some of the values here

Reply:
I get
~~~
ddtrace/internal/wrapping.py:416: error: Incompatible types in assignment (expression has type "Optional[MethodType]", variable has type "Optional[FunctionType]")  [assignment]
~~~
😢